### PR TITLE
[cominterop] Attach to runtime from IUnknown and IDispatch methods

### PR DESCRIPTION
**Pull request by lambdageek**
If the COM object for a managed object is passed to a thread that hasn't interacted with Mono before, we need to attach before doing GC thread state transitions.

---

Also added tests that invoke managed methods via a COM interface from an unattached thread.  This works (because the COM vtable entry is a native-to-managed wrapper which does a thread attach) but I don't think we had a test case before.

---

Addresses #18137

### CODE REVIEW

**Comment by lambdageek** [comment]
> Detach right away isn’t great but probably no good answer.

Ah yea, good point. I should mention that this is the same style of attach/detach that the native-to-managed wrappers do in [`emit_managed_wrapper_ilgen`](https://github.com/mono/mono/blob/master/mono/metadata/marshal-ilgen.c#L6054): it just does a GC thread state transition and unsets the domain.

There are multiple degrees of detach in the runtime (there's the managed thread object, the jit tls data, the GC thread state machine, etc) that destroy various degrees of state that the runtime has about each thread.

File: mono/tests/libtest.c [hunk 1]
```diff
@@ -7881,6 +7881,72 @@ mono_test_cominterop_ccw_queryinterface (MonoComObject *pUnk)
...
+#else
+	pthread_t t;
+	ccw_qi_shared_data *shared = malloc (sizeof (ccw_qi_shared_data));
+	shared->pUnk = pUnk;
+	shared->i = 1;
+	int res = pthread_create (&t, NULL, ccw_qi_foreign_thread, (void*)shared);
+	g_assert (res == 0);
+	pthread_join (t, NULL);
+	int result = shared->i;
+	g_free (shared);
```
Comment (lambdageek) [inline mono/tests/libtest.c]:
```suggestion
	free (shared);
```

File: mono/tests/libtest.c [hunk 2]
```diff
@@ -7881,6 +7881,72 @@ mono_test_cominterop_ccw_queryinterface (MonoComObject *pUnk)
...
+#else
+	pthread_t t;
+	ccw_qi_shared_data *shared = malloc (sizeof (ccw_qi_shared_data));
+	shared->pUnk = pUnk;
+	shared->i = 1;
+	int res = pthread_create (&t, NULL, ccw_itest_foreign_thread, (void*)shared);
+	g_assert (res == 0);
+	pthread_join (t, NULL);
+	int result = shared->i;
+	g_free (shared);
```
Comment (lambdageek) [inline mono/tests/libtest.c]:
```suggestion
	free (shared);
```

File: mono/tests/libtest.c [hunk 3]
```diff
@@ -7881,6 +7881,72 @@ mono_test_cominterop_ccw_queryinterface (MonoComObject *pUnk)
...
+}
+
+LIBTEST_API int STDCALL
+mono_test_cominterop_ccw_queryinterface_foreign_thread (MonoComObject *pUnk)
+{
+#ifdef WIN32
+	return 0;
+#else
+	pthread_t t;
+	ccw_qi_shared_data *shared = malloc (sizeof (ccw_qi_shared_data));
```
Comment (jaykrell) [inline mono/tests/libtest.c]: cast, i.e. for C++?

Comment (jaykrell) [reply mono/tests/libtest.c]: Check for out of memory, since this isn't `g_malloc`?

File: mono/tests/libtest.c [hunk 4]
```diff
@@ -7881,6 +7881,72 @@ mono_test_cominterop_ccw_queryinterface (MonoComObject *pUnk)
...
+}
+
+LIBTEST_API int STDCALL
+mono_test_cominterop_ccw_itest_foreign_thread (MonoComObject *pUnk)
+{
+#ifdef WIN32
+	return 0;
+#else
+	pthread_t t;
+	ccw_qi_shared_data *shared = malloc (sizeof (ccw_qi_shared_data));
```
Comment (jaykrell) [inline mono/tests/libtest.c]: ditto -- cast and check for out of memory?

**Comment by lambdageek** [comment]
@monojenkins backport to 2019-12

**Comment by lambdageek** [comment]
@monojenkins backport to 2019-10

**Pull request merged**